### ISSUE TITLE
Local function (types.FunctionType) serialization with Dask serializes entire package

### ISSUE THREAD

**Issue by devin-petersohn** [dask/distributed#3508]
I want to document a process I recently went through to solve a big serialization issue I ran into with Modin on Dask and how I fixed it. The fix is not good, and still serializes the entire package.

A succinct list of the high level causes of the issue:
* We have a library of functions that we define internally in the package.
* We often pass functions/other callables to a function we deploy as an argument (example below)
* We initialize Dask for users if it's not yet initialized when the library is imported
* Dask initialized without any custom serialization

### The issue

The issue that we ran into is that Dask will try to serialize the entire package when all I want to `client.submit` is some local function or lambda defined in an internal class. This results in Dask trying to create a bunch of new `Client` objects again because the `__init__.py` is being rerun on deserialization. I understand the motivation for serializing everything. However, if I use cloudpickle alone, it won't serialize everything, just the function and its dependencies.

### The fix
Right now I'm serializing the function arguments before they are passed as arguments to the function called in `client.submit`. Example:

```python
import cloudpickle as pkl

def apply_list_of_funcs(funcs, df):
        for func, kwargs in funcs:
            if isinstance(func, bytes):
                func = pkl.loads(func)
        df = func(df, **kwargs)
    return df

def apply(func, kwargs): # apply some function
    func = pkl.dumps(func)
    client.submit(apply_list_of_funcs, [(func, kwargs)], df)
```

Dask is still serializing the whole package and I get this warning because of that:
```
UserWarning: Large object of size 1.05 MB detected in task graph: 
  ([[<function BasePandasFrame._apply_index_objs.<lo ... 26cd6beac9b6a')
Consider scattering large objects ahead of time
...
```

#### Register custom serializer for `types.FunctionType`
I tried the following:

```python
import cloudpickle as pkl

@dask_serialize.register(types.FunctionType)
def serialize(func: types.FunctionType):
    header = {'serializer': 'cloudpickle'}
    frames = [pkl.dumps(func)]
    return header, frames


@dask_deserialize.register(types.FunctionType)
def deserialize(header, frames) -> types.FunctionType:
    if len(frames) > 1:
        frame = ''.join(frames)
    else:
        frame = frames[0]
    return pkl.loads(frame)
```

But this has no effect.

#### Change pickle to only use cloudpickle and initialize the client with `serializers=["pickle"]` and `deserializers=["pickle"]`

I also tried the following, thinking that the Dask serializer was causing the problem:
```python
import cloudpickle as pickle

def pickle_dumps(x):
    header = {'serializer': 'pickle'}
    frames = [pickle.dumps(x)]
    return header, frames


def pickle_loads(header, frames):
    if len(frames) > 1:  # this may be cut up for network reasons
        frame = ''.join(frames)
    else:
        frame = frames[0]
    return pickle.loads(frame)

from distributed.protocol.serialize import register_serialization_family
register_serialization_family('pickle', pickle_dumps, pickle_loads)
client = Client(serializers=["pickle"], deserializers=["pickle"])
```

This also had no effect.

### Conclusion

In short, I would like to know if there's a way to avoid this issue of serializing the entire package. I hope this writeup helps with diagnosis and toward a fix. Thanks!

**Comment by mrocklin** [comment]
> I understand the motivation for serializing everything. However, if I use cloudpickle alone, it won't serialize everything, just the function and its dependencies.

That's surprising.  We mostly just use cloudpickle by default.  This sort of thing usually happens because the function/method in question depends on some global state, which causes cloudpickle to serialize the world.  Usually the solution is to make these state dependencies explicit in the function and pass in things as keyword arguments (which I imagine you already know).

You could try using `distributed.protocol.serialize` on your method and see if that serializes the world.  If it does, then maybe take another look at cloudpickle?

**Comment by devin-petersohn** [comment]
I took a break and thought about this for a bit, I had gotten a bit too deep and went down a rabbit hole that led me to asking the wrong question. Putting aside the "serialize the world" problem for a moment, the problem we are having is the serializer is creating a bunch of new `Client`s when `__init__.py` is run after deserialization. With cloudpickle, we have some really strange workaround that seems to work in `__init__.py`:

```python
if threading.current_thread().name == "MainThread":
    client = Client()
```

This works when I manually use cloudpickle on the function. This isn't working by default. 

The fix I wrote above is just passing raw bytes of the function created by `cloudpickle.dumps` as an argument to the function, then in the `client.submit` function deserializing with `cloudpickle.loads`. This will not try to create a bunch of new clients with the workaround (read vile hack) I showed above because the condition evaluated to `False`. This could be solved by not serializing everything, but for the moment I would like to understand if there's some way to see why manually pickling is different from submitting the function as an argument. Even `distributed.protocol.serialize` is using pickle, and looking through the Dask source I do not see anything that would cause the behaviors to be different.

**Comment by mrocklin** [comment]
We just use pickle/cloudpickle.  We're a bit clever about which one we use and when (pickle is faster in many cases), but at the end it's just that.

It sounds like you have a `Client()` object inside of a module, which seems off.  I recommend finding a way to avoid that if possible.  Client objects like that have LocalClusters inside of them, which have processes.  Having that much system specific state in a module is asking for lots of trouble.

I would recommend that you have users create Dask clients on their own, and use the `get_client()` function to get the default client that they have created.  It's atypical for libraries to manage their own Dask clusters.

**Comment by devin-petersohn** [comment]
Thanks @mrocklin I will consider it. We already allow users to create their own `Client` and we connect to it if it exists. If not we create one for them.

There are certainly challenges to creating a `Client`, but our users aren't Dask users and typically they are unfamiliar with Dask/multiprocessing/whatever, so abstracting this away for them allows them to not worry about getting the parameters right or even have to think about it.

I will chalk the serialization thing up to something weird I'm doing. It seems like that is your conclusion as well.

Closing as resolved, thanks for the help!

**Comment by mrocklin** [comment]
If you're going to create a client and have it sit in a module then you
might want to try setting processes=False.  That may smooth things over for
you and reduce a lot of the machinery involved.
